### DISCUSSION TITLE
NDI support in Air

### DISCUSSION THREAD

**Discussion by ramraver** [airsdk/Adobe-Runtime-Support#2113] [General]
Does someone know if NDI support is planned in Air?

**Comment by ajwfrost** [comment]
That's the kind of thing we hope to enable with the new multimedia pipeline -> basically to extend the ANE interfaces so that you could write an ANE as a kind of plug-in into the media pipeline, and then be able to display the output via the AIR display list and existing output sinks such as `Video`, `VideoTexture`, `StageVideo`...

I'm a little unclear with these folk though in terms of exactly how it may work; the idea of "royalty free" implies they're not supporting H.264, HEVC etc... -> will go have a chat with them at IBC at the end of next week!

thanks

**Reply by ramraver** [reply]
Thanks. Please keep us informed about this. 
Another huge benefit would be support for a video format enabling use of transparency/alpha channel in air. There once was the on2vp6 before it was taken out.

**Comment by ajwfrost** [comment]
Update here: we still have some doubts about licensing terms, and looking at the NDI agreements it's unlikely that we'd be able to built support directly into AIR. However with the changes that we're planning for the multimedia extensibility via ANEs, then it would be possible to create an ANE that links with the NDI framework client libraries and means that video content from NDI can be viewed within AIR, controlled by the AS3 code and displayed via the normal display options..

**Reply by teotigraphix** [reply]
So it would be possible to do a round trip, consuming a video signal through NDI, "mixing it", then outputting as a target for another NDI enabled software?

What would the performance be in your opinion? I use VJ software and can see uses if it actually worked decent.

**Reply by ajwfrost** [reply]
I'm not sure about the "mixing it" part ... we would have video sources - which could be coming from NDI - and video sinks - which are the AS3 Video, VideoTexture, StageVideo (and maybe NativeWindow) objects. Mixing videos together isn't implicit in any of these...

There was talk about having a way to present the overall AIR output as a video feed (maybe via NDI?) which was requested for integration with things like OBS. That's a separate topic though.. happy to look into it but it changes the whole usage idea of AIR as an application framework into something a lot more powerful!

Performance-wise for all of this -> NDI takes care of decoding the software and can even do colour space conversion, I hope it uses hardware and/or SIMD accelerations for this, so it should be okay. I wasn't massively impressed by the demos that we get from the NDI tools/samples but then having seen demos at IBC, I suspect it can be done without too much of a challenge. The issue though is then in composition when we receive the data from NDI; if we're CPU-rendering then it wouldn't be a massive impact I think, but if we're having to upload every decoded frame into graphics memory for GPU composition, it could be a bit slower; better approach would be to use native OS decoders that keep the decoded frames available in e.g. DirectX...

thanks
